### ISSUE TITLE
Cannot `nix develop` - encountering errors.

### ISSUE THREAD

**Issue by wafflepool-cardano** [geniusyield/smart-order-router#26]
System: Ubuntu 22.04 LTS

- git cloned the repository. Installed Nix as per the guideline: https://github.com/input-output-hk/plutus-apps/blob/main/CONTRIBUTING.adoc#installing-and-setting-up-nix (inlcuding appending a line in /etc/nix/nix.conf)

- go to project root directory. `nix develop` - enter yes, yes, yes.

Error:
`warning: Using saved setting for 'allow-import-from-derivation = true' from ~/.local/share/nix/trusted-settings.json.
warning: Using saved setting for 'extra-substituters = https://cache.iog.io' from ~/.local/share/nix/trusted-settings.json.
warning: Using saved setting for 'extra-trusted-public-keys = hydra.iohk.io:f/Ea+s+dFdN+3Y/G+FDgSq+a5NEWhJGzdjvKNGv0/EQ=' from ~/.local/share/nix/trusted-settings.json.
warning: ignoring untrusted substituter 'https://cache.iog.io'
error: input 'haskellNix/hydra/nixpkgs' follows a non-existent input 'nix/nixpkgs'
       … while updating the lock file of flake 'git+file:///home/user/smart-order-router?ref=main&rev=ca466980214be8d40bafa381ad69e27cddb8198d'
`

Any clue as to how to solve this? 
As well, is there any discord or telegram channel where testers of the smart order router can discuss and share their progress?

Many thanks.

**Comment by sourabhxyz** [comment]
> Any clue as to how to solve this?

@alegadea Could you update nix instructions to explain how to handle for these. Probably related to setting `nix.conf` file.

Fwiw, if you have your environment setup as in this CI file https://github.com/geniusyield/atlas/blob/main/.github/workflows/haskell.yml, then you can build & run without nix.

> As well, is there any discord or telegram channel where testers of the smart order router can discuss and share their progress?

Unfortunately, no. But as this request has come up many times, I'll raise it internally.

**Comment by wafflepool-cardano** [comment]
I copied the haskell.yml file to smart-order-router/.github/workflow/haskell.yml.

Getting the following error when running `cabal build all`:
```
Warning: Requested index-state 2023-06-06T00:00:00Z is newer than
'hackage.haskell.org'! Falling back to older state (2023-06-05T23:38:20Z).
Resolving dependencies...
Error: cabal: Could not resolve dependencies:
[__0] next goal: atlas-cardano (user goal)
[__0] rejecting: atlas-cardano-0.3.0 (conflict: requires GHC2021 which is not
supported)
[__0] fail (backjumping, conflict set: atlas-cardano)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: atlas-cardano
```
Is there something in specific I need to modify for this to work? 

Noted, I will keep looking out for such a channel on the Genius Yield discord.

**Comment by sourabhxyz** [comment]
Hi, I didn't mean to copy `haskell.yml` file, I meant that if you are not using nix, then you would need to make your current environment suitable. I.e., you would need appropriate versions of `GHC`, `cabal`, `libsodium`, `secp256k1`, etc. installed. You can see which versions of `GHC` (which is `9.2.8`), `cabal` (which is `3.10.1.0`), etc. we use by looking through the linked `haskell.yml` file.

Basically, your environment should be able to build cardano-node from source (see [this](https://developers.cardano.org/docs/get-started/installing-cardano-node/) guide) and if you are able to build it from source, just change version of GHC & cabal as mentioned for in this message.

**Comment by wafflepool-cardano** [comment]
Good morning, 

Regarding setting up Kupo as a local provider in `atlas-config-kupo.json`, what is the expected input for `"kupoUrl": "<<KUPO_URL>>"` ?
I'm looking through the documentation but unsure of which format to provide and whether I should provide an address as well.

![Screenshot from 2023-10-12 12-08-46](https://github.com/geniusyield/smart-order-router/assets/80946800/93798008-7f1b-4c71-ba91-1b183acfacb7)


Please advise.

**Comment by sourabhxyz** [comment]
`http://localhost:1442`.

**Comment by wafflepool-cardano** [comment]
Thank you.

So i configured everything, including the bot file. Also implemented kupo url as per your advice. Now receiving the following error when i `make orderbot-kupo` in the root directory of the project:

![Screenshot from 2023-10-12 16-49-04](https://github.com/geniusyield/smart-order-router/assets/80946800/f138d949-8eb0-4709-82a6-de0f5f132048)

Port 1442 is opened to localhost (127.0.0.1). Path to node socket has been triple checked and is correct.

Any idea of what might be going on?

**Comment by wafflepool-cardano** [comment]
**For more context:**
config-files/atlas-config-kupo.json
`{
  "coreProvider": {
    "socketPath": "/opt/cardano/cnode/sockets/node0.socket",
    "kupoUrl": "http://localhost:1442"
  },
  "networkId": "testnet-preprod",
  "logging": [
    {
      "type": {
        "tag": "stderr"
      },
      "severity": "Info",
      "verbosity": "V2"
    }
  ]
}
`
bot-config.json:
`{ "signingKeyFP" : "config-files/wallet/payment1.skey"
, "collateral": "ed3f13a307a1fbfe7b0eb6deda5dbedba5c5d79e80876827c0d3f8cda9ef9570#0"

, "nftMintingPolicyFP" : "compiled-scripts/minting-policy"
, "orderValidatorFP" : "compiled-scripts/partial-order"
, "validatorRefs":
  { "refAddr": "addr_test1wpgexmeunzsykesf42d4eqet5yvzeap6trjnflxqtkcf66g0kpnxt"
  , "refNftAC": "fae686ea8f21d567841d703dea4d4221c2af071a6f2b433ff07c0af2.e6a295bb83d06f53fcf91151f54acec0a63fbd6f0d924206d5d012e6da3b72af"
  , "refNftUtxoRef": "39f987a6beb9cc4c45bba149a21c28068f640f3593f15f8157f0b6022b431977#0"
  , "scriptRef": "39f987a6beb9cc4c45bba149a21c28068f640f3593f15f8157f0b6022b431977#1"
  , "nftPolicyRef": "39f987a6beb9cc4c45bba149a21c28068f640f3593f15f8157f0b6022b431977#0"
  }

, "strategy" : "OneSellToManyBuy"
, "scanDelay" : 40000000
, "maxOrderMatches" : 5
, "maxTxsPerIteration" : 5
, "randomizeMatchesFound": true

, "scanTokens" :
 [
    { "commodityAsset" : "c6e65ba7878b2f8ea0ad39287d3e2fd256dc5c4160fc19bdf4c4d87e.7447454e53"
    , "currencyAsset" : "lovelace"
    }
  ]
}
`

**Comment by sourabhxyz** [comment]
Could you also share the command you used to run kupo?

**Comment by wafflepool-cardano** [comment]
Oh I didn't realise kupo had to be run separately.

```
kupo \
--node-socket /opt/cardano/cnode/sockets/node0.socket \
--node-config /opt/cardano/cnode/files/config.json \
--since origin \
--match \
--workdir /home/user/kupo/db \
```
What is the recommended value for `--match`? Is it necessary to sync since origin?

**Comment by sourabhxyz** [comment]
This is what I use:

```
kupo \
	--node-socket $HOME/Cardano/testnet/db/node.socket \
	--node-config $HOME/Cardano/node-8.1.2/preprod-configs/config.json \
	--since origin \
	--match "*" \
	--prune-utxo \
	--workdir $HOME/Cardano/kupo/db
```
> What is the recommended value for --match?

You can go with `*` though it's an overkill for this use case but it doesn't hurt either.

> Is it necessary to sync since origin?

No, depends upon things like how old is your bot address etc., but can go with origin like I do.

**Comment by wafflepool-cardano** [comment]
Thank you for your prompt reply.

Kupo is on and syncing. Will try the to `make orderbot-kupo` once it's done. 
However, doesn't having `--match *` use up a lot of the computer's memory?

**Comment by sourabhxyz** [comment]
Persistent disk storage -> ~1.3 gb for Preprod & 12.6 gb for Mainnet. RAM usage is almost non-existent.

**Comment by sourabhxyz** [comment]
Also, if you are running kupo for the first time, also add `--defer-db-indexes` flag (see https://cardanosolutions.github.io/kupo/#section/Getting-started) to speed up syncing and in second run, you can omit for it.

Fwiw, one is not bound to kupo as we also have Maestro provider option.

**Comment by wafflepool-cardano** [comment]
Seems like it is working now:
![Screenshot from 2023-10-16 15-30-20](https://github.com/geniusyield/smart-order-router/assets/80946800/b78c93f9-6838-48bb-b746-26c2e1457b41)

Does the `unsafeAddressFromText` flag indicate anything negative?
![Screenshot from 2023-10-16 15-32-41](https://github.com/geniusyield/smart-order-router/assets/80946800/efd14f26-ab5c-437b-891f-e5116d66062d)

**Comment by Micrograx** [comment]
Hi, I'm glad everything is working now!

The `unsafeAddressFromText` is just part of the Atlas log that shows which addresses are being queried. Nothing to be worried about

**Comment by wafflepool-cardano** [comment]
Great, thank you :)
I will leave it running as is then. Anything else I should be looking out for in terms of monitoring? As well, any update on a dedicated discord (or other) channel for the smart order routers?

**Comment by sourabhxyz** [comment]
Hi @wafflepool-cardano, I would like to mention that running Kupo on `*` match pattern would quickly burn out SSDs TBW! Please see the updated README on how to run Kupo efficiently and a sample script is provided [here](https://github.com/geniusyield/smart-order-router/tree/main/scripts).

Also, now we have a dedicated discord channel! See #discussion-sors in GeniusYield discord.
